### ISSUE TITLE
Update with where clossure is not working as expected

### ISSUE THREAD

**Issue by michalbundyra** [laminas/laminas-db#89]
In Zend\Db\TableGateway\AbstractTableGateway update method if you send where as clossure is sent as a parameter in update->where
`if ($where !== null) {
            $update->where($where);
        }
`
So the closure end up being called with a PredicateInterface object but in the select and delete method
the code is
`if ($where instanceof \Closure) {
            $where($delete);
        } else {
            $delete->where($where);
        }
`

`if ($where instanceof \Closure) {
            $where($select);
        } elseif ($where !== null) {
            $select->where($where);
        }
`

And there the closure is called with a SqlInterface object and I can't understand why the update method is not the same as select and delete ? Because if you want to make a generic clossure for select,update and delete as I made the it  won't work because of the different interface object passed in closure.

---

Originally posted by @WADS94 at https://github.com/zendframework/zend-db/issues/183

**Comment by michalbundyra** [comment]
Tl;Dr, what is your usecase of reusing closure?

Played with this for a while. At first appeared as careless implementation needing a fix. Did the fix in code, first in favor of passing PredicateSet to closure to give opportunity to modify it using a reusable closure for convenience across all 3 operation types. But when got to updating documentation on closure usage, realized it gives opportunity to fully customize `Select` object in ways beyond what TableGateway offers out of the box. So I tried to inverse the parameter passing full `Update` and `Delete` into closure. This caused a worse dilemma. It makes good sense to modify `Select` in more details beyond just `where`, such as `order by`, `limit` etc. But `Update` and `Delete` will want whole different usage. In fact, allowing only `PredicateInterface` is a pretty good restriction to have. Then, there is interface difference. `SqlInterface` is pretty limited, so having it serve all 3 different objects (or 2 different categories, to be more exact - one for detailed querying and another for restricted changes).

Usage does look ugly but I am not seeing how to improve it.

---

Originally posted by @alextech at https://github.com/zendframework/zend-db/issues/183#issuecomment-264728502

**Comment by weierophinney** [comment]
> This package is considered feature-complete, and is now in **security-only** maintenance mode, following a [decision by the Technical Steering Committee](https://github.com/laminas/technical-steering-committee/blob/2b55453e172a1b8c9c4c212be7cf7e7a58b9352c/meetings/minutes/2020-08-03-TSC-Minutes.md#vote-on-components-to-mark-as-security-only).
> If you have a security issue, please [follow our security reporting guidelines](https://getlaminas.org/security/).
> If you wish to take on the role of maintainer, please [nominate yourself](https://github.com/laminas/technical-steering-committee/issues/new?assignees=&labels=Nomination&template=Maintainer_Nomination.md&title=%5BNOMINATION%5D%5BMAINTAINER%5D%3A+%7Bname+of+person+being+nominated%7D)
>
> If you are looking for an actively maintained package alternative, we recommend:
>
> - [doctrine/dbal](https://www.doctrine-project.org/projects/dbal.html)
